dali/core/mm/mm_test_utils.h

```diff
@@ -29,18 +30,24 @@ namespace mm {
 namespace test {
 
 template <bool owning, bool security_check,
-          typename Base, typename Upstream, typename... ExtraParams>
-class test_resource_wrapper : public Base {
+          typename Base, typename Upstream>
+class test_resource_wrapper;
+
+template <bool owning, bool security_check, typename Upstream>
+class test_resource_wrapper_impl {
+ protected:
   using sentinel_t = uint32_t;
-  void *do_allocate(ExtraParams... extra, size_t bytes, size_t alignment) override {
+
+  template <typename UpstreamAlloc, typename... Extra>
+  void *do_allocate_impl(UpstreamAlloc &&ua, size_t bytes, size_t alignment, Extra&&... extra) {
     if (simulate_oom_)
       throw std::bad_alloc();
     void *ret;
     if (security_check) {
-      ret = upstream_->allocate(extra..., bytes + sizeof(sentinel_t), alignment);
-      detail::write_sentinel<sentinel_t>(ret, bytes);
+      ret = ua(bytes + sizeof(sentinel_t), alignment);
+      detail::write_sentinel<sentinel_t>(ret, bytes, std::forward<Extra>(extra)...);
     } else {
-      ret = upstream_->allocate(extra..., bytes, alignment);
+      ret = upstream_->allocate(bytes, alignment, std::forward<Extra>(extra)...);
```

Comment:
Why use `ua` in one place and `upstream_` directly in the second?

Reply:
It's a bug. It should be `ua`.